kivy/lang.py

```diff
@@ -1910,6 +1910,10 @@ def _apply_rule(self, widget, rule, rootrule, template_ctx=None):
         for crule in rule.children:
             cname = crule.name
 
+            if cname in ('canvas', 'canvas.before', 'canvas.after'):
+                raise Exception('canvas instructions added in kv must '
```

Comment:
This should use `ParserException`; i.e. `ParserException(crule.ctx, crule.line, 'canvas instructions....')`. If you change that then :+1: for merging!